modules/ml/src/main/java/org/apache/ignite/ml/dataset/UpstreamTransformer.java

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.dataset;
+
+import java.io.Serializable;
+import java.util.Random;
+import java.util.stream.Stream;
+
+/**
+ * Class encapsulating transformer of upstream.
+ *
+ * @param <K> Type of keys in the upstream.
+ * @param <V> Type of values in the upstream.
+ * @param <T> Data needed for this transformer.
+ */
+public abstract class UpstreamTransformer<K, V, T> implements Serializable {
+    /**
+     * Performs transformation.
+     *
+     * @param rnd RNG.
+     * @param upstream Upstream.
+     * @return Transformed upstream.
+     */
+    public final Stream<UpstreamEntry<K, V>> transform(Random rnd, Stream<UpstreamEntry<K, V>> upstream) {
```

Comment:
It would be cleaner this way if we only needed specifying seeds in the moment of transformer creation, but we need to change seed for random on partitions, so it is much easier to have one point of randomness control within the chain which controls all randomness of transformers instead of somehow transferring needed seeds to those transformers which need it.